src/modules/lost/functions.c

```diff
@@ -698,70 +1055,128 @@ int lost_function(struct sip_msg *_m, char *_con, char *_uri, char *_name,
 		}
-		LM_INFO("### LOST uri\t[%.*s]\n", uri.len, uri.s);
-		/* get the displayName element */
-		name.s = lost_get_content(root, name_element, &name.len);
-		if(name.len == 0) {
-			LM_ERR("displayName element not found: [%.*s]\n", ret.len, ret.s);
-			/* free memory */
-			lost_free_string(&ret);
-			goto err;
-		}
-		LM_INFO("### LOST din\t[%.*s]\n", name.len, name.s);
-	} else if((!xmlStrcmp(root->name, (const xmlChar *)"errors"))) {
-
-		LM_DBG("findService error response received\n");
-
-		/* get the error patterm */
-		err.s = lost_get_childname(root, errors_element, &err.len);
-		LM_DBG("findService error response: [%.*s]\n", err.len, err.s);
-		if(err.len == 0) {
-			LM_ERR("error pattern element not found: [%.*s]\n", ret.len, ret.s);
-			/* free memory */
-			lost_free_string(&ret);
-			goto err;
+		switch(fsrdata->category) {
+			case RESPONSE:
+				if(fsrdata->uri != NULL) {
+					/* get the first uri element */
+					if((tmp.s = fsrdata->uri->value) != NULL) {
+						tmp.len = strlen(fsrdata->uri->value);
+						uri.s = lost_copy_string(tmp, &uri.len);
+					}
+				} else {
+					LM_ERR("uri not found: [%.*s]\n", ret.len, ret.s);
+					goto err;
+				}
+				if(fsrdata->mapping != NULL) {
+					/* get the displayName element */
+					if((tmp.s = fsrdata->mapping->name->text) != NULL) {
+						tmp.len = strlen(fsrdata->mapping->name->text);
+						name.s = lost_copy_string(tmp, &name.len);
+					}
+				} else {
+					LM_ERR("name not found: [%.*s]\n", ret.len, ret.s);
+					goto err;
+				}
+				/* we are done */
+				redirect = 0;
+				break;
+			case ERROR:
+				/* get the errors element */
+				if(fsrdata->errors != NULL) {
+					if((tmp.s = fsrdata->errors->issue->type) != NULL) {
+						tmp.len = strlen(fsrdata->errors->issue->type);
+						err.s = lost_copy_string(tmp, &err.len);
+					}
+					/* clean up */
+					tmp.s = NULL;
+					tmp.len = 0;
+				} else {
+					LM_ERR("errors not found: [%.*s]\n", ret.len, ret.s);
+					goto err;
+				}
+				/* we are done */
+				redirect = 0;
+				break;
+			case REDIRECT:
+				/* get the target element */
+				if(fsrdata->redirect != NULL) {
+					if((tmp.s = fsrdata->redirect->target) != NULL) {
+						tmp.len = strlen(fsrdata->redirect->target);
+						url.s = &(ustr[0]);
+						url.len = MAX_URI_SIZE;
+						if((naptr = lost_naptr_lookup(tmp, &shttps, &url))
+								== 0) {
+							naptr = lost_naptr_lookup(tmp, &shttp, &url);
+						}
+						if(naptr == 0) {
+							LM_ERR("NAPTR failed on [%.*s]\n", tmp.len, tmp.s);
+							goto err;
+						}
+						/* clean up */
+						tmp.s = NULL;
+						tmp.len = 0;
+						/* check loop */
+						if(oldurl.s != NULL && oldurl.len > 0) {
+							if(str_strcasecmp(&url, &oldurl) == 0) {
+								LM_ERR("loop detected: "
+									   "[%.*s]<-->[%.*s]\n",
+										oldurl.len, oldurl.s, url.len, url.s);
+								goto err;
+							}
+						}
+						/* remember the redirect target */
+						oldurl.s = lost_copy_string(url, &oldurl.len);
```

Comment:
if you need a str anyway, you can use pkg_str_dup

Reply:
thanks, have incorporated it